src/main/java/com/bisnode/opa/client/rest/JsonBodyHandler.java

```diff
@@ -0,0 +1,47 @@
+package com.bisnode.opa.client.rest;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.FieldDefaults;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.http.HttpResponse;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+@RequiredArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+public class JsonBodyHandler<T> implements HttpResponse.BodyHandler<Supplier<T>> {
+
+    JavaType javaType;
+
+    public static <W> HttpResponse.BodySubscriber<Supplier<W>> asJSON(JavaType javaType) {
+        HttpResponse.BodySubscriber<InputStream> upstream = HttpResponse.BodySubscribers.ofInputStream();
+
+        return HttpResponse.BodySubscribers.mapping(upstream, createMapper(javaType));
+    }
+
+    private static <W> Function<InputStream, Supplier<W>> createMapper(JavaType javaType) {
+        return is -> () -> mapToJson(javaType, is);
+    }
+
+    private static <W> W mapToJson(JavaType javaType, InputStream is) {
+        try (InputStream stream = is) {
```

Comment:
Please rename `stream` to `inputStream`. When you look at line 37 without context it's easy to get confused.